src/core/thread/mle_router.cpp

```diff
@@ -414,68 +412,51 @@ void MleRouter::SetRouterDowngradeThreshold(uint8_t aThreshold)
     mRouterDowngradeThreshold = aThreshold;
 }
 
-void MleRouter::HandleAdvertiseTimer(void *aContext)
+bool MleRouter::HandleRouterAdvertiseTimer(void *aContext)
 {
-    MleRouter *obj = reinterpret_cast<MleRouter *>(aContext);
-    obj->HandleAdvertiseTimer();
+    MleRouter *obj = static_cast<MleRouter *>(aContext);
+    return obj->HandleAdvertiseTimer();
 }
 
-void MleRouter::HandleAdvertiseTimer(void)
+void MleRouter::HandleReedAdvertiseTimer(void *aContext)
 {
-    uint32_t advertiseDelay;
+    MleRouter *obj = static_cast<MleRouter *>(aContext);
 
-    if ((mDeviceMode & ModeTlv::kModeFFD) == 0)
+    if (obj->HandleAdvertiseTimer())
     {
-        return;
+        uint32_t advertiseDelay = Timer::SecToMsec(kReedAdvertiseInterval +
+                                                   (otPlatRandomGet() % kReedAdvertiseJitter));
+        obj->mReedAdvertiseTimer.Start(advertiseDelay);
     }
+}
 
-    SendAdvertisement();
-
-    switch (GetDeviceState())
+bool MleRouter::HandleAdvertiseTimer(void)
+{
+    if ((mDeviceMode & ModeTlv::kModeFFD) == 0)
     {
-    case kDeviceStateDisabled:
-    case kDeviceStateDetached:
-        assert(false);
-        break;
-
-    case kDeviceStateChild:
-        advertiseDelay = Timer::SecToMsec(kReedAdvertiseInterval +
-                                          (otPlatRandomGet() % kReedAdvertiseJitter));
-        mAdvertiseTimer.Start(advertiseDelay);
-        break;
-
-    case kDeviceStateRouter:
-    case kDeviceStateLeader:
-        mAdvertiseInterval *= 2;
+        return false;
+    }
 
-        if (mAdvertiseInterval > kAdvertiseIntervalMax)
-        {
-            mAdvertiseInterval = kAdvertiseIntervalMax;
-        }
+    SendAdvertisement();
 
-        advertiseDelay = Timer::SecToMsec(mAdvertiseInterval) / 2;
-        advertiseDelay += otPlatRandomGet() % (advertiseDelay);
-        mAdvertiseTimer.Start(advertiseDelay);
-        break;
-    }
+    return true;
 }
 
-ThreadError MleRouter::ResetAdvertiseInterval(void)
+void MleRouter::ResetAdvertiseInterval(void)
 {
-    uint32_t advertiseDelay;
-
-    VerifyOrExit(mAdvertiseInterval != kAdvertiseIntervalMin || !mAdvertiseTimer.IsRunning(), ;);
+    assert(GetDeviceState() == kDeviceStateRouter ||
```

Comment:
This assert seems to be getting hit everytime in the tests. Why would this be failing. I thought REEDs shouldn't be using this because they only use the reed timing variables.